src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -1,54 +1,67 @@
-effect operation functions (that is, functions that use 'send') from a given
-effect algebra. For example, using the @FileSystem@ effect from the example in
-the module documentation for "Polysemy", we can write the following:
-
-@
-data FileSystem m a where
-  ReadFile :: 'FilePath' -> FileSystem 'String'
-  WriteFile :: 'FilePath' -> 'String' -> FileSystem ()
-'makeSem' ''FileSystem
-@
-
-This will automatically generate the following functions:
+{-# LANGUAGE NamedFieldPuns  #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TupleSections   #-}
 
-@
-readFile :: 'Member' FileSystem r => 'FilePath' -> 'Sem' r 'String'
-readFile a = 'send' (ReadFile a)
-
-writeFile :: 'Member' FileSystem r => 'FilePath' -> 'String' -> 'Sem' r ()
-writeFile a b = 'send' (WriteFile a b)
-@
--}
+-- | This module provides Template Haskell functions for automatically generating
+-- effect operation functions (that is, functions that use 'send') from a given
+-- effect algebra. For example, using the @FileSystem@ effect from the example in
+-- the module documentation for "Polysemy", we can write the following:
+--
+-- @
+-- data FileSystem m a where
+--   ReadFile  :: 'FilePath' -> FileSystem 'String'
+--   WriteFile :: 'FilePath' -> 'String' -> FileSystem ()
+--
+-- 'makeSem' ''FileSystem
+-- @
+--
+-- This will automatically generate (approximately) the following functions:
+--
+-- @
+-- readFile :: 'Member' FileSystem r => 'FilePath' -> 'Sem' r 'String'
+-- readFile a = 'send' (ReadFile a)
+--
+-- writeFile :: 'Member' FileSystem r => 'FilePath' -> 'String' -> 'Sem' r ()
+-- writeFile a b = 'send' (WriteFile a b)
+-- @
 module Polysemy.Internal.TH.Effect
   ( makeSem
   , makeSem_
-  )
-where
+  ) where
+
+import Prelude hiding ((<>))
 
-import Control.Monad (join, forM, unless)
-import Data.Char (toLower)
+import Control.Monad
+import Data.Bifunctor
+import Data.Char                      (toLower)
+import Data.Either
+import Data.Generics
 import Data.List
-import Generics.SYB
+import Data.Tuple
 import Language.Haskell.TH
-import Polysemy.Internal (send, Member, Sem)
+import Language.Haskell.TH.PprLib
+import Language.Haskell.TH.Datatype
+import Polysemy.Internal              (send, Member, Sem)
 import Polysemy.Internal.CustomErrors (DefiningModule)
 
+import qualified Data.Map.Strict as M
 
--- | If @T@ is a GADT representing an effect algebra, as described in the module
--- documentation for "Polysemy", @$('makeSem' ''T)@ automatically
+-- TODO: write tests for what should (not) compile
+
+------------------------------------------------------------------------------
+-- | If @T@ is a datatype representing an effect algebra, as described in the
+-- module documentation for "Polysemy", @$('makeSem' ''T)@ automatically
 -- generates a smart constructor for every data constructor of @T@.
 --
+-- Datatype may be created with GADT syntax, traditional syntax and can be
```

Comment:
I think I'd prefer to not publish this fact; life is easier for everyone if we all agree just to use GADT syntax.

Reply:
Maybe just have a comment "this also will work for data families", so that people who have a use case won't be scared off, but people who don't won't be inspired to try to make things more complicated.

Reply:
Done.